### PR TITLE
add remark components

### PR DESCRIPTION
**Pull request by josemarin7**
https://github.com/iterative/blog/issues/16 
I've tried to resolve this issue with gatsby-remark-component.
https://www.gatsbyjs.org/packages/gatsby-remark-component

### CODE REVIEW

**Comment by shcheklein** [comment]
@josemarin7 is it supposed to be merged this way? I see some commented imports, some test files (test.md)?

File: content/blog/test.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  This is description post
+author: ../authors/svetlana_grinchenko.md
+commentsUrl: https://discuss.dvc.org/t/november-19-dvc-heartbeat/284
+tags:
+  - Test
+---
+
+This is content post
+
+<div>
```
Comment (shcheklein) [inline content/blog/test.md]: can we don w/o div?

File: src/components/external-link/index.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  description,
+  picture
+}: IExternalLinkProps) => {
+  return (
+    <a href={href} className="external-link-preview">
+      <section className="elp-content-holder">
+        <div className="elp-description-holder">
+          <h4 className="elp-title">{title}</h4>
+          <div className="elp-description">{description}</div>
+          <div className="elp-link">deploymachinelearning.com</div>
```
Comment (shcheklein) [inline src/components/external-link/index.tsx]: is it always the same link?

File: src/components/external-link/index.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        </div>
+        <div className="elp-image-holder">
+          <img src={picture} style={{ maxWidth: 700 }} />
```
Comment (shcheklein) [inline src/components/external-link/index.tsx]: should we move styles to a proper place?

File: src/components/post/index.tsx [hunk 4]
```diff
@@ -7,17 +7,18 @@ import { useWindowScroll, useWindowSize } from 'react-use';
 import { IBlogPostData } from '../../templates/blog-post';
 
 import { getCommentsCount } from '../../api';
-import { siteLinks } from '../../data';
+// import { siteLinks } from '../../data';
```
Comment (shcheklein) [inline src/components/post/index.tsx]: dead code?

File: src/templates/blog-post.tsx [hunk 5]
```diff
@@ -80,6 +81,7 @@ export const pageQuery = graphql`
       id
       excerpt(format: HTML)
       html
```
Comment (shcheklein) [inline src/templates/blog-post.tsx]: is html still used?

File: src/components/post/index.tsx [hunk 6]
```diff
@@ -100,9 +101,8 @@ function Post({ html, timeToRead, frontmatter, fields }: IBlogPostData) {
           )}
         </>
       )}
-      <div className={styles.content}>
-        <Markdown html={html} />
-      </div>
+      <div className={styles.content}>{renderAst(htmlAst)}</div>
+      {/*<Markdown html={html} />*/}
```
Comment (shcheklein) [inline src/components/post/index.tsx]: does it mean that Markdown is not needed anymore?

**Comment by iAdramelk** [comment]
@josemarin7 @shcheklein we are now adding a lot of code to the client (e. g. `rehype-react`, `unist`, `bluebird` and it's dependencies) and we don't really need all this stuff at runtime. Can we move the node transformation to the server, do it in `gatsby-node` and leave the client code as is?

File: src/templates/blog-post.tsx (hunk 5 above)
Comment (josemarin7) [reply src/templates/blog-post.tsx]: No, html is not using anymore.  Please see https://www.gatsbyjs.org/packages/gatsby-remark-component guide

File: src/components/post/index.tsx (hunk 6 above)
Comment (josemarin7) [reply src/components/post/index.tsx]: Yes. we need to use htmlAst instead of html. we can rewrite Markdown component for htmlAst

**Comment by josemarin7** [comment]
@iAdramelk I've followed step by step guid in official plugin doc https://www.gatsbyjs.org/packages/gatsby-remark-component 
I've just installed gatsby-remark-component and rehype-react packages.

In fact, I don't understand latest comment.
`Can we move the node transformation to the server, do it in gatsby-node and leave the client code as is?`

File: content/blog/test.md (hunk 1 above)
Comment (josemarin7) [reply content/blog/test.md]: All below are correct syntax:
```
<external-link
  href="https://veekaybee.github.io/2019/02/13/data-science-is-different/"
  title="Deploy Machine Learning Models with Django"
  description="Version 1.0 (04/11/2019) Piotr Płoński The demand for Machine Learning (ML) applications is growing. Many resources..."
  picture="/uploads/images/2019-12-14/how-to-manage-your-machine-learning-workflow.jpeg" />
```
```<external-link
  href="https://veekaybee.github.io/2019/02/13/data-science-is-different/"
  title="Deploy Machine Learning Models with Django"
  description="Version 1.0 (04/11/2019) Piotr Płoński The demand for Machine Learning (ML) applications is growing. Many resources..."
  picture="/uploads/images/2019-12-14/how-to-manage-your-machine-learning-workflow.jpeg">
</external-link>
```
```
<external-link
  href="https://veekaybee.github.io/2019/02/13/data-science-is-different/"
  title="Deploy Machine Learning Models with Django"
  description="Version 1.0 (04/11/2019) Piotr Płoński The demand for Machine Learning (ML) applications is growing. Many resources..."
  picture="/uploads/images/2019-12-14/how-to-manage-your-machine-learning-workflow.jpeg"></external-link>
```

File: src/components/external-link/index.tsx (hunk 2 above)
Comment (josemarin7) [reply src/components/external-link/index.tsx]: Fixed, see utils/getHostname

File: src/components/external-link/index.tsx (hunk 3 above)
Comment (josemarin7) [reply src/components/external-link/index.tsx]: I think my task was to find a way to insert react components into MD files. I do not know what styles should be here

File: src/components/markdown/index.tsx [hunk 7]
```diff
@@ -14,5 +18,17 @@ function Markdown({ html }: IMarkdownProps) {
     />
   );
 }
+*/
+
+function Markdown({ html, htmlAst }: IMarkdownProps) {
+  return htmlAst ? (
```
Comment (shcheklein) [inline src/components/markdown/index.tsx]: why do we need to support both?

what is the performance impact of rendering this client-side?

File: src/utils/getHostname.ts [hunk 8]
```diff
@@ -0,0 +1,4 @@
+export const getHostname = (url: string = '') => {
```
Comment (shcheklein) [inline src/utils/getHostname.ts]: is there a standard JS function that can do this? don't like the idea of supporting our custom regex

File: content/blog/test.md [hunk 9]
```diff
@@ -0,0 +1,18 @@
+---
```
Comment (shcheklein) [inline content/blog/test.md]: do we actually want to merge this file? transform one of the existing documents?

File: src/components/post/index.tsx [hunk 10]
```diff
@@ -7,8 +7,8 @@ import { useWindowScroll, useWindowSize } from 'react-use';
 import { IBlogPostData } from '../../templates/blog-post';
 
 import { getCommentsCount } from '../../api';
-import { siteLinks } from '../../data';
 import { pluralizeComments } from '../../utils/i18n';
+// import { renderAst } from '../../utils/renderAst';
```
Comment (shcheklein) [inline src/components/post/index.tsx]: dead code?

File: content/blog/test.md (hunk 1 above)
Comment (josemarin7) [reply content/blog/test.md]: div removed

File: src/components/post/index.tsx (hunk 6 above)
Comment (josemarin7) [reply src/components/post/index.tsx]: I've rewrite Markdown comp to accept htmlAst prop

File: content/blog/test.md (hunk 9 above)
Comment (josemarin7) [reply content/blog/test.md]: No, it is just for test. Once there will be any issue on my PR, then I will remove this test.md

**Comment by iAdramelk** [comment]
> @iAdramelk I've followed step by step guid in official plugin doc https://www.gatsbyjs.org/packages/gatsby-remark-component
> I've just installed gatsby-remark-component and rehype-react packages.
> 
> In fact, I don't understand latest comment.
> `Can we move the node transformation to the server, do it in gatsby-node and leave the client code as is?`

Hi, sorry for the sort comment. The idea is this:

Right now, all remark parsing and generating is happening on the server in the build step. We just send the resulting HTML to the browser. This way, we don't need to add JavaScript libraries to the JS bundle and spend time to parse markdown and generate HTML on the client. It's a good thing because the smaller the bundle, the faster the page is loading, and the less JS-code to parse and execute, the faster the page works.

Example in `gatsby-remark-component` is a working one, but:

1. It adds additional dependencies like `rehype-react` to the bundle.
2. It replaces HTML with htmlAST, and htmlAST is slightly larger (arguably after gzip difference will be negligible, but still).
3. Using `rehype-react` and transforming its nodes will make the page render slightly slower.

It's not a critical thing, but if we have an easy way to prevent it will be better. And we have such a way:

Gatsby gives us access to the markdown nodes in the build step, and we can read them and modify them at the build time. We can add hook [onCreateNode](https://www.gatsbyjs.org/docs/node-apis/#onCreateNode) to the `gatsby-node.js` file and change our markdown (using `remark`) or HTML source (using `rehype`) there on the server instead of doing it in the browser.

**Pull request closed without merging**